File: bot.py
"""
WhatsApp MCQ Bot - Main bot logic
"""

import json
import time
import random
from datetime import datetime, timedelta
from typing import Dict, List, Optional
from waha import WAHAClient
from config import (
    QUESTIONS_CHANNEL_ID, 
    ANSWERS_CHANNEL_ID, 
    EMOJI_OPTIONS, 
    BOT_NAME,
    DAILY_QUESTIONS
)

class MCQBot:
    """Main bot class for handling MCQ questions and answers"""
    
    def __init__(self):
        self.waha_client = WAHAClient()
        self.questions = self.load_questions()
        self.sent_questions = self.load_sent_questions()
        
    def load_questions(self) -> List[Dict]:
        """Load questions from JSON file"""
        try:
            with open('questions.json', 'r', encoding='utf-8') as f:
                data = json.load(f)
                return data.get('questions', [])
        except FileNotFoundError:
            print("❌ questions.json not found")
            return []
        except json.JSONDecodeError as e:
            print(f"❌ Error parsing questions.json: {e}")
            return []
    
    def load_sent_questions(self) -> List[int]:
        """Load list of already sent question IDs"""
        try:
            with open('sent_questions.json', 'r') as f:
                return json.load(f)
        except FileNotFoundError:
            return []
    
    def save_sent_questions(self):
        """Save list of sent question IDs"""
        with open('sent_questions.json', 'w') as f:
            json.dump(self.sent_questions, f)
    
    def get_random_question(self) -> Optional[Dict]:
        """Get a random question that hasn't been sent recently"""
        available_questions = [q for q in self.questions if q['id'] not in self.sent_questions]
        
        if not available_questions:
            # Reset sent questions if all have been used
            self.sent_questions = []
            available_questions = self.questions
        
        if available_questions:
            question = random.choice(available_questions)
            self.sent_questions.append(question['id'])
            self.save_sent_questions()
            return question
        
        return None
    
    def format_question_message(self, question: Dict, question_num: int) -> str:
        """Format question message for WhatsApp"""
        message = f"📚 Daily MCQ Challenge - Question {question_num}/{DAILY_QUESTIONS}\n\n"
        message += f"{question['question']}\n\n"
        
        # Add options with emojis
        for option, text in question['options'].items():
            emoji = EMOJI_OPTIONS.get(option, "❓")
            message += f"{option}) {text} {emoji}\n"
        
        message += f"\nReact with the emoji of your answer!\n"
        message += f"⏰ Answer will be revealed in 5 minutes\n"
        message += f"📊 Category: {question.get('category', 'General')}\n"
        message += f"🎯 Difficulty: {question.get('difficulty', 'Medium')}"
        
        return message
    
    def format_answer_message(self, question: Dict, question_num: int, reactions: Dict = None) -> str:
        """Format answer message for WhatsApp"""
        correct_option = question['correct_answer']
        correct_text = question['options'][correct_option]
        correct_emoji = EMOJI_OPTIONS.get(correct_option, "✅")
        
        message = f"📚 Daily MCQ Challenge - Answer {question_num}/{DAILY_QUESTIONS}\n\n"
        message += f"{question['question']}\n\n"
        message += f"✅ Correct Answer: {correct_option}) {correct_text} {correct_emoji}\n\n"
        message += f"💡 Explanation: {question['explanation']}\n\n"
        
        # Add reaction statistics if available
        if reactions:
            message += "📊 Vote Results:\n"
            total_reactions = sum(reactions.values())
            
            for option, text in question['options'].items():
                emoji = EMOJI_OPTIONS.get(option, "❓")
                count = reactions.get(emoji, 0)
                percentage = (count / total_reactions * 100) if total_reactions > 0 else 0
                marker = "✅" if option == correct_option else ""
                message += f"{emoji} {option}) {text} - {count} votes ({percentage:.1f}%) {marker}\n"
            
            message += f"\nTotal: {total_reactions} participants"
        else:
            message += "📊 Vote results will be available after reactions are counted"
        
        return message
    
    def get_message_reactions(self, message_id: str) -> Dict:
        """Get reactions for a specific message"""
        try:
            reactions = self.waha_client.get_reactions(message_id)
            if "reactions" in reactions:
                return reactions["reactions"]
            return {}
        except Exception as e:
            print(f"Error getting reactions: {e}")
            return {}
    
    def send_question(self, question_num: int) -> bool:
        """Send a question to the questions channel"""
        question = self.get_random_question()
        if not question:
            print("❌ No questions available")
            return False
        
        message = self.format_question_message(question, question_num)
        
        print(f"📤 Sending question {question_num} to Questions Channel...")
        result = self.waha_client.send_message(QUESTIONS_CHANNEL_ID, message)
        
        if "error" in result:
            print(f"❌ Failed to send question: {result['error']}")
            return False
        
        print(f"✅ Question {question_num} sent successfully")
        print(f"Question: {question['question']}")
        print(f"Correct Answer: {question['correct_answer']}) {question['options'][question['correct_answer']]}")
        
        return True
    
    def send_answer(self, question_num: int) -> bool:
        """Send answer to the answers channel"""
        # Get the last sent question
        if not self.sent_questions:
            print("❌ No questions have been sent yet")
            return False
        
        last_question_id = self.sent_questions[-1]
        question = next((q for q in self.questions if q['id'] == last_question_id), None)
        
        if not question:
            print(f"❌ Question with ID {last_question_id} not found")
            return False
        
        message = self.format_answer_message(question, question_num)
        
        print(f"📤 Sending answer {question_num} to Answers Channel...")
        result = self.waha_client.send_message(ANSWERS_CHANNEL_ID, message)
        
        if "error" in result:
            print(f"❌ Failed to send answer: {result['error']}")
            return False
        
        print(f"✅ Answer {question_num} sent successfully")
        return True
    
    def send_question_and_answer(self, question_num: int) -> bool:
        """Send both question and answer simultaneously"""
        print(f"\n🚀 Sending Question {question_num}/{DAILY_QUESTIONS}")
        print("=" * 50)
        
        # Send question
        question_sent = self.send_question(question_num)
        if not question_sent:
            return False
        
        # Wait a moment
        time.sleep(2)
        
        # Send answer
        answer_sent = self.send_answer(question_num)
        if not answer_sent:
            return False
        
        print(f"✅ Question {question_num} and answer sent successfully!")
        return True
    
    def send_daily_questions(self) -> bool:
        """Send all daily questions"""
        print(f"🤖 {BOT_NAME} - Starting daily question session")
        print(f"📅 Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        print("=" * 60)
        
        success_count = 0
        
        for i in range(1, DAILY_QUESTIONS + 1):
            if self.send_question_and_answer(i):
                success_count += 1
            
            # Wait between questions (except for the last one)
            if i < DAILY_QUESTIONS:
                print(f"⏳ Waiting 10 seconds before next question...")
                time.sleep(10)
        
        print("=" * 60)
        print(f"📊 Daily session completed: {success_count}/{DAILY_QUESTIONS} questions sent")
        
        if success_count == DAILY_QUESTIONS:
            print("🎉 All questions sent successfully!")
            return True
        else:
            print("⚠️ Some questions failed to send")
            return False
    
    def test_connection(self) -> bool:
        """Test WhatsApp connection"""
        print("🔧 Testing WhatsApp connection...")
        
        status = self.waha_client.get_session_status()
        if status.get("status") == "WORKING":
            print("✅ WhatsApp connection is working")
            return True
        else:
            print(f"❌ WhatsApp connection status: {status.get('status', 'Unknown')}")
            return False

def main():
    """Main function"""
    bot = MCQBot()
    
    # Test connection
    if not bot.test_connection():
        print("❌ WhatsApp connection failed. Please check your setup.")
        return
    
    # Send daily questions
    bot.send_daily_questions()

if __name__ == "__main__":
    main()

File: bot.js
/**
 * WhatsApp MCQ Bot - Main bot logic using Baileys
 */

const BaileysClient = require('./baileys-client');
const fs = require('fs');
const path = require('path');

class MCQBot {
    constructor() {
        this.baileysClient = new BaileysClient();
        this.questions = this.loadQuestions();
        this.sentQuestions = this.loadSentQuestions();
        
        // Configuration
        this.questionsChannelId = process.env.QUESTIONS_CHANNEL_ID;
        this.answersChannelId = process.env.ANSWERS_CHANNEL_ID;
        this.dailyQuestions = 4;
        this.botName = 'MCQ Bot';
        
        // Emoji mappings
        this.emojiOptions = {
            'A': '😀',
            'B': '😁',
            'C': '😂',
            'D': '🤣'
        };
        
        if (!this.questionsChannelId || !this.answersChannelId) {
            throw new Error('QUESTIONS_CHANNEL_ID and ANSWERS_CHANNEL_ID environment variables are required');
        }
    }

    loadQuestions() {
        try {
            const data = fs.readFileSync('questions.json', 'utf8');
            const parsed = JSON.parse(data);
            return parsed.questions || [];
        } catch (error) {
            console.error('❌ Error loading questions:', error.message);
            return [];
        }
    }

    loadSentQuestions() {
        try {
            const data = fs.readFileSync('sent_questions.json', 'utf8');
            return JSON.parse(data);
        } catch (error) {
            return [];
        }
    }

    saveSentQuestions() {
        try {
            fs.writeFileSync('sent_questions.json', JSON.stringify(this.sentQuestions, null, 2));
        } catch (error) {
            console.error('❌ Error saving sent questions:', error.message);
        }
    }

    getRandomQuestion() {
        const availableQuestions = this.questions.filter(q => !this.sentQuestions.includes(q.id));
        
        if (availableQuestions.length === 0) {
            // Reset sent questions if all have been used
            this.sentQuestions = [];
            return this.questions[Math.floor(Math.random() * this.questions.length)];
        }
        
        const question = availableQuestions[Math.floor(Math.random() * availableQuestions.length)];
        this.sentQuestions.push(question.id);
        this.saveSentQuestions();
        return question;
    }

    formatQuestionMessage(question, questionNum) {
        let message = `📚 Daily MCQ Challenge - Question ${questionNum}/${this.dailyQuestions}\n\n`;
        message += `${question.question}\n\n`;
        
        // Add options with emojis
        for (const [option, text] of Object.entries(question.options)) {
            const emoji = this.emojiOptions[option] || '❓';
            message += `${option}) ${text} ${emoji}\n`;
        }
        
        message += `\nReact with the emoji of your answer!\n`;
        message += `⏰ Answer will be revealed in 5 minutes\n`;
        message += `📊 Category: ${question.category || 'General'}\n`;
        message += `🎯 Difficulty: ${question.difficulty || 'Medium'}`;
        
        return message;
    }

    formatAnswerMessage(question, questionNum) {
        const correctOption = question.correct_answer;
        const correctText = question.options[correctOption];
        const correctEmoji = this.emojiOptions[correctOption] || '✅';
        
        let message = `📚 Daily MCQ Challenge - Answer ${questionNum}/${this.dailyQuestions}\n\n`;
        message += `${question.question}\n\n`;
        message += `✅ Correct Answer: ${correctOption}) ${correctText} ${correctEmoji}\n\n`;
        message += `💡 Explanation: ${question.explanation}\n\n`;
        message += `📊 Vote results will be available after reactions are counted`;
        
        return message;
    }

    async sendQuestion(questionNum) {
        const question = this.getRandomQuestion();
        if (!question) {
            console.log('❌ No questions available');
            return false;
        }

        const message = this.formatQuestionMessage(question, questionNum);
        
        console.log(`📤 Sending question ${questionNum} to Questions Channel...`);
        try {
            await this.baileysClient.sendMessageToChannel(this.questionsChannelId, message);
            console.log(`✅ Question ${questionNum} sent successfully`);
            console.log(`Question: ${question.question}`);
            console.log(`Correct Answer: ${question.correct_answer}) ${question.options[question.correct_answer]}`);
            return true;
        } catch (error) {
            console.error(`❌ Failed to send question: ${error.message}`);
            return false;
        }
    }

    async sendAnswer(questionNum) {
        if (this.sentQuestions.length === 0) {
            console.log('❌ No questions have been sent yet');
            return false;
        }

        const lastQuestionId = this.sentQuestions[this.sentQuestions.length - 1];
        const question = this.questions.find(q => q.id === lastQuestionId);
        
        if (!question) {
            console.log(`❌ Question with ID ${lastQuestionId} not found`);
            return false;
        }

        const message = this.formatAnswerMessage(question, questionNum);
        
        console.log(`📤 Sending answer ${questionNum} to Answers Channel...`);
        try {
            await this.baileysClient.sendMessageToChannel(this.answersChannelId, message);
            console.log(`✅ Answer ${questionNum} sent successfully`);
            return true;
        } catch (error) {
            console.error(`❌ Failed to send answer: ${error.message}`);
            return false;
        }
    }

    async sendQuestionAndAnswer(questionNum) {
        console.log(`\n🚀 Sending Question ${questionNum}/${this.dailyQuestions}`);
        console.log('='.repeat(50));
        
        // Send question
        const questionSent = await this.sendQuestion(questionNum);
        if (!questionSent) {
            return false;
        }
        
        // Wait a moment
        await new Promise(resolve => setTimeout(resolve, 2000));
        
        // Send answer
        const answerSent = await this.sendAnswer(questionNum);
        if (!answerSent) {
            return false;
        }
        
        console.log(`✅ Question ${questionNum} and answer sent successfully!`);
        return true;
    }

    async sendDailyQuestions() {
        console.log(`🤖 ${this.botName} - Starting daily question session`);
        console.log(`📅 Date: ${new Date().toLocaleString()}`);
        console.log('='.repeat(60));
        
        let successCount = 0;
        
        for (let i = 1; i <= this.dailyQuestions; i++) {
            if (await this.sendQuestionAndAnswer(i)) {
                successCount++;
            }
            
            // Wait between questions (except for the last one)
            if (i < this.dailyQuestions) {
                console.log('⏳ Waiting 10 seconds before next question...');
                await new Promise(resolve => setTimeout(resolve, 10000));
            }
        }
        
        console.log('='.repeat(60));
        console.log(`📊 Daily session completed: ${successCount}/${this.dailyQuestions} questions sent`);
        
        if (successCount === this.dailyQuestions) {
            console.log('🎉 All questions sent successfully!');
            return true;
        } else {
            console.log('⚠️ Some questions failed to send');
            return false;
        }
    }

    async testConnection() {
        console.log('🔧 Testing WhatsApp connection...');
        return this.baileysClient.isReady();
    }
}

async function main() {
    try {
        const bot = new MCQBot();
        
        // Connect to WhatsApp
        console.log('🔗 Connecting to WhatsApp...');
        try {
            await bot.baileysClient.connect();
        } catch (error) {
            console.log('⚠️ WhatsApp connection attempt completed');
        }
        
        // Test connection
        if (!await bot.testConnection()) {
            console.log('❌ WhatsApp not connected yet.');
            console.log('📱 Please scan the QR code with your WhatsApp app');
            console.log('==========================================');
            console.log('📱 FOR MOBILE USERS:');
            console.log('1. Copy the QR URL from above');
            console.log('2. Go to: https://www.qr-code-generator.com/');
            console.log('3. Paste the URL to generate QR code');
            console.log('4. Scan with WhatsApp');
            console.log('==========================================');
            console.log('💻 FOR DESKTOP USERS:');
            console.log('🖼️  QR Code image saved as: whatsapp-qr.png');
            console.log('🌐 Open qr-display.html in a browser');
            console.log('==========================================');
            console.log('🔄 Run the workflow again after scanning to send messages');
            
            // Exit gracefully
            process.exit(0);
        }
        
        // Send daily questions
        await bot.sendDailyQuestions();
        
        // Disconnect
        await bot.baileysClient.disconnect();
        
    } catch (error) {
        console.error('❌ Bot execution failed:', error.message);
        process.exit(1);
    }
}

// Run the bot
if (require.main === module) {
    main();
}

module.exports = MCQBot;